keps/79-hierarchical-cohorts/README.md

```diff
@@ -0,0 +1,242 @@
+
+With this proposal, research org's top Cohort will simply set borrowingLimit to 0. Alternatively, production
+org's top Cohort can set lendingLimit to 0. BorrowingLimitis inside production org's ClusterQueues should
+be generous enough to allow borrowing from the research org.
+
+#### Story 2
+
+I have a couple organizations that have dedicated resources. The organizations should not borrow 
+from each other, however I want to have an additional "special" queue, with low priority jobs, that can 
+borrow unused capacity from any of the organizations.
+
+With this proposal, the cohorts for organizations will set borrowingLimit to 0. Top level Cohort will 
+contain all of these Cohorts, plus the "special" ClusterQueue, with borrowingLimit set to infinity. 
+
+### Risks and Mitigations
+
+* Users may create a cycle in the Cohort hierarchy - Kueue will stop all new admissions within
+the entire tree. The already admitted workloads will be allowed to continue. Appropriate 
+ClusterQueue/Cohort Status Conditions will be set and Events emited.
+
+* Scheduling and preemption may require more computation/resources.
+
+## Design Details
+
+The Cohort API will initially start only with the basic functionality. Additional policies
+regarding sharing Cohort resources can be added later.
+
+```go
+
+type Cohort struct {
+    metav1.TypeMeta   `json:",inline"`
+    metav1.ObjectMeta `json:"metadata,omitempty"`
+
+    Spec   CohortSpec   `json:"spec,omitempty"`
+    Status CohortStatus `json:"status,omitempty"`
+}
+
+type CohortSpec struct {
+    // Cohort parent name. The parent Cohort object doesn't have to exist.
+    // In such case, it is assumed that parent simply doesn't have any
+    // quota and limits and doesn't have any other custom settings.
+    Parent *string `json:"parent,omitempty"`
+
+    // resourceGroups describes groups of resources that the Cohort can
+    // share with ClusterQueues within the same group of Cohorts/ClusterQueues.
+    // Each resource group defines the list of resources and a list of flavors
+    // that provide quotas for these resources.
+    // Each resource and each flavor can only form part of one resource group.
+    // resourceGroups can be up to 16.
+    //
+    // BorrowingLimit specifies how much ClusterQueues under this Cohort can borrow
+    // from ClusterQueues/Cohorts that are NOT under this Cohort. For Cohorts without
+    // a parent(top of the hierarchy) the BorrowingLimit has to be 0.
+    //
+    // LendingLimit specifies how much ClusterQueues that are NOT under this Cohort
+    // can borrow from the ClusterQueues/Cohorts that are under this Cohort.
+    // If any of the Limits is not specified it means that there is no limit
+    // and ClusterQueues can borrow/lend as much as they want/have.
+    // 
+    // +listType=atomic
+    // +kubebuilder:validation:MaxItems=16
+    ResourceGroups []ResourceGroup `json:"resourceGroups,omitempty"`
+}
+
+
+type CohortStatus struct {
```

Comment:
Do we need a filed to show the resource borrowed from others?

Reply:
This may be problematic as with multi-level hierarchy, a single admission may trigger status updates on many levels, increasing the need for etcd write throughput, which remains one of the biggest Kueue bottlenecks. 

I would consider doing the detailed stats for Cohorts via dedicated api, something similar to: https://github.com/kubernetes-sigs/kueue/blob/main/apis/visibility/v1alpha1/types.go

Reply:
Please leave a comment about this in the doc.

Reply:
Done.